### PR TITLE
ENH: metadata PEP390 setup.cfg

### PR DESCRIPTION
**Pull request by scivision**
Helping modernize Meson packaging

* Begin modernizing Meson install metadata per PEP 390 [using `setup.cfg` with setup.py](https://setuptools.readthedocs.io/en/latest/setuptools.html#configuring-setup-using-setup-cfg-files) (available since Dec 2016).
* Moved flake8 content to dedicated [standard .flake8 file](https://flake8.pycqa.org/en/latest/user/configuration.html#configuration-locations)
* added [mypy.ini](https://mypy.readthedocs.io/en/latest/config_file.html) for some good defaults as Meson continues to introduce type hint checks

setup.py can be reduced to 4 lines with future PRs--I thought I'd introduce gradually with "safe" metadata fields only in case of unexpected issues. 
I use setup.cfg on hundreds of repos with setup.py a one-liner: `import setuptools; setuptools.setup()`

---

Future: [Poetry](https://pypi.org/project/poetry/) is growing ~ 250K installs/month. May be nice to add pyproject.toml suitable for Poetry 

---

mlog.py: allowed ipython kernel to be used for debugging in say Spyder. In any case, having a universal fallback to non-colorized in case of unexpected issue seems good.

run_project_test.py: cleaned up syntax w.r.t. file checks by using pathlib.Path

### CODE REVIEW

**Comment by jpakkane** [comment]
```
File .\.eggs\README.txt contains DOS line ending on line 1. Only unix-style line endings are permitted.
```

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    # E305: expected 2 blank lines after class or function definition, found 1
+    E305,
+    # E401: multiple imports on one line
+    E401,
+    # E266: too many leading '#' for block comment
+    E266,
+    # E402: module level import not at top of file
+    E402,
+    # E731: do not assign a lambda expression, use a def (too many false positives)
+    E731
```
Comment (QuLogic) [inline .flake8]: Where are the commas after this (and subsequent) lines?

File: setup.cfg [hunk 2]
```diff
@@ -1,33 +1,35 @@
...
-    # W504: line break after binary operator
-    W504
-    # A003: builtin class attribute
-    A003
-max-line-length = 120
+[metadata]
+description = A high performance build system
+author = Jussi Pakkanen
+author_email = jpakkane@gmail.com
+url = http://mesonbuild.com
```
Comment (QuLogic) [inline setup.cfg]:
```suggestion
url = https://mesonbuild.com
```

File: .flake8 (hunk 1 above)
Comment (scivision) [reply .flake8]: added

Comment (scivision) [reply .flake8]: added

**Comment by jpakkane** [comment]
You need to edit `run_project_tests.py` so that `check_format()` ignores files in directories called `.eggs`.

**Comment by scivision** [comment]
With regard to https://github.com/mesonbuild/meson/pull/5592/commits/28df6cfeae4ed7994bde379150d7b560d0c0ac31#diff-8f34612f804ea9e4e0e9ea36b32e4361R22 from #5592, [Python 3.5.2 is required](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING)

**Comment by jpakkane** [comment]
Seems like a temporary glitch:

```
Download error on https://pypi.org/simple/: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056) -- Some packages may not be found!
No local packages or working download links found for pip>=10
```

Restarted the job.

**Comment by scivision** [comment]
Re the pip glitch, it appears that was just a transient error as that is passing now. Appears to be just the usual -- all tests pass but azure gets return code of 1

**Comment by scivision** [comment]
rebased on master to check that Azure Cygwin errors reduced

**Comment by jpakkane** [comment]
Failure seems sporadic, so I restarted the job:

```
test_meson_installed (__main__.CommandTests) ... Download error on https://pypi.org/simple/pip/: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1076) -- Some packages may not be found!
Couldn't find index page for 'pip' (maybe misspelled?)
Download error on https://pypi.org/simple/: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1076) -- Some packages may not be found!
No local packages or working download links found for pip>=10
```

**Comment by scivision** [comment]
I just restarted the failed jobs now. Maybe some Azure connectivity issues; for example before I reran I saw on `vs2019 vc2019x64ninja`:

```
##[section]Starting: PowerShell
==============================================================================
Task         : PowerShell
Description  : Run a PowerShell script on Linux, macOS, or Windows
Version      : 2.151.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
==============================================================================
Generating script.
========================== Starting Command Output ===========================
##[command]"C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'd:\a\_temp\d990cdfe-4c0b-45fc-a5d0-babff7d39c67.ps1'"
DMD32 D Compiler v2.087.0

Copyright (C) 1999-2019 by The D Language Foundation, All Rights Reserved written by Walter Bright
dub.exe : Package urld not found for registry at https://code.dlang.org/ (fallbacks registry at 
https://code-mirror.dlang.io/, registry at https://dub-registry.herokuapp.com/): HTTP request returned status code 502 
(Bad Gateway)
At D:\a\_temp\d990cdfe-4c0b-45fc-a5d0-babff7d39c67.ps1:84 char:6
+      & dub fetch urld
+      ~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (Package urld no...2 (Bad Gateway):String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
##[error]PowerShell exited with code '1'.
##[section]Finishing: PowerShell
```

**Comment by jpakkane** [comment]
The problem [seems to be in setup_requires](https://github.com/pypa/pip/issues/410).

**Comment by scivision** [comment]
Yes that was it, those setup_requires weren't necessary in this case because if one had older version than them, setup.cfg isn't read anyway, so they really weren't doing anything.

**Pull request merged**